src/wallet/wallet2.cpp

```diff
@@ -6090,6 +6100,19 @@ crypto::chacha_key wallet2::get_ringdb_key()
   return *m_ringdb_key;
 }
 
+void wallet2::register_devices(){
+  hw::trezor::register_all();
+}
+
+hw::device& wallet2::lookup_device(const std::string & device_descriptor){
+  if (!m_devices_registered){
+    m_devices_registered = true;
+    register_devices();
+  }
+
+  return hw::get_device(device_descriptor);
+}
+
```

Comment:
Why is this needed here now ?

Reply:
The `hw::device& wallet2::lookup_device(const std::string & device_descriptor)` was added in order to break cyclic dependency device_trezor and wallet. 

Device_trezor uses a wallet function so I had to define wallet shim

```c++
typedef struct wallet_shim {
    boost::function<crypto::public_key (const tools::wallet2::transfer_details &td)> get_tx_pub_key_from_received_outs;
  } wallet_shim;
```

Shim prevents linking dependency device_trezor on wallet. Because of this the device registration is performed in runtime - reason why `lookup_device` was added and why wallet links to `device_trezor`, so it can call `hw::trezor::register_all();`.